From 695db3e1b125220cf3fbe670c5686675f24b612d Mon Sep 17 00:00:00 2001
Message-ID: <695db3e1b125220cf3fbe670c5686675f24b612d.1779975966.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Fri, 22 May 2026 10:55:53 +0200
Subject: [PATCH 09/30] virtio: Recheck transport IOMMU feature after pre-plug

Transport pre-plug hooks can add host features before the virtio device is
fully plugged. In particular, transports may add VIRTIO_F_IOMMU_PLATFORM
before virtio_bus_device_plugged() selects the DMA address space.

Recheck the feature after transport hooks and feature population so
transport-provided DMA address spaces can be selected when needed.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-bus.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/hw/virtio/virtio-bus.c b/hw/virtio/virtio-bus.c
index cef944e01588..f8d2127e70ed 100644
--- a/hw/virtio/virtio-bus.c
+++ b/hw/virtio/virtio-bus.c
@@ -75,6 +75,13 @@ void virtio_bus_device_plugged(VirtIODevice *vdev, Error **errp)
         return;
     }
 
+    /*
+     * pre_plugged()/get_features may add transport-provided DMA features
+     * (for example VIRTIO_F_IOMMU_PLATFORM). Re-evaluate the flag after
+     * feature population so get_dma_as() can be engaged when needed.
+     */
+    has_iommu |= virtio_host_has_feature(vdev, VIRTIO_F_IOMMU_PLATFORM);
+
     if (klass->device_plugged != NULL) {
         klass->device_plugged(qbus->parent, &local_err);
     }
-- 
2.53.0

